hashicorp/nomad
# cni: use check command when restoring from restart

When the Nomad client restarts and restores allocations, the network namespace for an allocation may exist but no longer be correctly configured. For example, if the host is rebooted and the task was a Docker task using a pause container, the network namespace may be recreated by the docker daemon.

When we restore an allocation, use the CNI "check" command to verify that any existing network namespace matches the expected configuration. This requires CNI plugins of at least version 1.2.0 to avoid a bug in older plugin versions that would cause the check to fail. If the check fails, fail the restore so that the allocation can be recreated (rather than silently having networking fail).

This should fix the gap left #24650 for Docker task drivers and any other drivers with the `MustInitiateNetwork` capability.

Fixes: https://github.com/hashicorp/nomad/issues/24292
Ref: https://github.com/hashicorp/nomad/pull/24650

### Testing & Reproduction steps

Run a cluster on a set of VMs, with at least one client. This can't be a server+client because we need to reboot the hosts. You should probably set the server.heartbeat_grace = "5m" to give yourself time to work.

* Run a Docker/Podman task with network.mode = "bridge". Wait for it to be healthy.
* Restart the agent.
* Make sure the alloc is untouched and networking works.
* Reboot the client host.
* Make sure the alloc is restored, the tasks are restarted, and networking works.

### Contributor Checklist
  changelog entry using the `make cl` command.
  ensure regressions will be caught.
  and job configuration, please update the  Nomad website documentation to reflect this. Refer to
  the [website README](../website/README.md) for docs guidelines. Please also consider whether the
  change requires notes within the [upgrade guide](../website/content/docs/upgrade/upgrade-specific.mdx).

### Reviewer Checklist
  backporting document.
  in the majority of situations. The main exceptions are long-lived feature branches or merges where
  history should be preserved.
  within the public repository.
